### PR TITLE
Update ConnectionPool docs

### PR DESCRIPTION
**Pull request by nycdotnet**
### Motivation / Background

I created this Pull Request because it was not clear to me that Rails models would automatically use the connection checked-out by `.with_connection` without me explicitly passing the connection to them.  This led to me digging in to this a bit more.  I think the docs could have been more explicit on this point, so I attempted to improve them.

### Detail

This Pull Request changes the docs to be more explicit about the behavior of `.with_connection`, when connections will be checked out and checked back into the connection pool, and what happens if a connection has already been checked out on the current thread.

### Additional information

N/A

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] CI is passing.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -206,10 +212,15 @@ def release_connection(owner_thread = ActiveSupport::IsolatedExecutionState.cont
...
       end
 
-      # If a connection obtained through #connection or #with_connection methods
-      # already exists yield it to the block. If no such connection
-      # exists checkout a connection, yield it to the block, and checkin the
-      # connection when finished.
+      # Yields a connection from the connection pool to the block. If no connection
+      # is already checked out by the current thread, a connection will be checked
+      # out from the pool, yielded to the block, and then returned to the pool when
+      # the block is finished.  If a connection has already been checked out on the
```
Comment (p8) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:218]:
```suggestion
      # the block is finished. If a connection has already been checked out on the
```

**Comment by nycdotnet** [comment]
@p8 thanks for your suggestion.  I made another fix as well.

**Comment by p8** [comment]
Thanks @nycdotnet. I think this improves the documentation.
Can you squash your commits?

**Comment by nycdotnet** [comment]
@p8 this looks like it's a flaky build issue?

**Comment by nycdotnet** [comment]
good to know - ty!

**Pull request merged**